HLT-compatible LHCInfoPer* PopCon

JanChyczynski opened the pull request:
#### PR description:

Update of the duringFill mode of LHCInfoPer* O2Os to be able to write to HLT and express tags complying to the synchronization rules. It writes only one (or none) payload at execution and writes to an IOV corresponding to a lumisection not yet processed by HLT. It's achieved by using `OnlineDBOutputService::writeIOVForNextLumisection` instead of `PoolDBOutputService::writeMany`. The O2Os in duringFill now uses `OnlinePopCon` and `OnlinePopConAnalyzer` (implemented in #44927) instead of the original ones.

Depends on #44927 (has to be tested on top of it and merged after it)

There is a plan to add better unit tests (not necessarily in this PR).

#### PR validation:

Tested by running the workflow in various states of LHC (stable beams, no stable beams etc.) for a few fills. Each time the logs procured and the uploaded payloads were examined.
The unit tests for the `duringFill` mode have been updated (although its extension is planned).
Example commands for running the workflow:
```
cd CondTools/RunInfo/test/
cmsRun ../python/LHCInfoPerLSPopConAnalyzer_cfg.py mode=duringFill destinationConnection=sqlite:test.db \
lastLumiFile='last_lumi.txt'  tag=perls_test_tag startTime='2022-06-26 15:43:20.000'

cmsRun ../python/LHCInfoPerFillPopConAnalyzer_cfg.py mode=duringFill destinationConnection=sqlite:fill_during/test.db \
lastLumiFile='last_lumi.txt' tag=perfill_test_tag  startTime='2022-06-26 15:43:20.000'
```

#### Backport 
14_1_X backport: https://github.com/cms-sw/cmssw/pull/46572

JanChyczynski committed:
adjust LHCInfoPer*PopConSourceHandler duringFill mode for HLT-compatibility

adjust LHCInfoPerFillPopConSourceHandler duringFill mode for HLT-compatibility, minors for perLS

fix LHCInfoPerLS PopCon duringFill startSampleTime error

fix filtering in  duringFill mode (prevPayload error)

Remove adding empty payload to empty tag

JanChyczynski committed:
Make perLS and perFill duringFill use OnlinePopCon

Make perLS duringFill use OnlinePopCon

Make perFill duringFill use OnlinePopCon

Attempt to configure OnlineDBOutputService in LHCInfoPer*PopConAnalyzer_cfg.py

finish LHCInfoPer*PopConAnalyzer_cfg files for updated duringFill

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"1e801ace29477dbb42621cac17285d67880f884f":{"files":["CondTools/RunInfo/interface/LHCInfoPerFillPopConSourceHandler.h","CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc","CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc","CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc","CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc"],"squashed":false,"time":1720373432},"346e1dcdf274211488f85920eacff33d28a7121e":{"files":["CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzer_cfg.py","CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzer_cfg.py"],"squashed":false,"time":1720373434},"428c47b557178aa4d9d19339c4e9e4d16edb58d0":{"files":["CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h","CondTools/RunInfo/plugins/BuildFile.xml","CondTools/RunInfo/plugins/LHCInfoPerLSOnlinePopConAnalyzer.cc","CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzer_cfg.py"],"squashed":false,"time":1720373431},"50107e6dbf056026d84c91f39cbef9193ced6c04":{"files":["CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc","CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc"],"squashed":false,"time":1720371465},"54289f9f5218327afa66d532e75ea04360896045":{"files":["CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h","CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc","CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc"],"squashed":false,"time":1720373434},"6805b11e2e5934456f62700a67f2143d06076d93":{"files":["CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc","CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzer_cfg.py"],"squashed":false,"time":1720371465},"78f902fb66c501032bc4d3fe29d6a0969e81c301":{"files":["CondTools/RunInfo/interface/LHCInfoPerLSPopConSourceHandler.h","CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc","CondTools/RunInfo/src/LHCInfoPerLSPopConSourceHandler.cc"],"squashed":false,"time":1720373430},"7e10c85909a44cf565c7ff0399729dba72a72c64":{"files":["CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc"],"squashed":false,"time":1720373427},"9907b9bb498a876c73914c2d34d9fa96202bd430":{"files":["CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc","CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc"],"squashed":false,"time":1720373429},"9deef0df69da1df74ee3af2e67f82b908f02a534":{"files":["CondTools/RunInfo/plugins/LHCInfoPerFillPopConAnalyzer.cc","CondTools/RunInfo/plugins/LHCInfoPerLSPopConAnalyzer.cc"],"squashed":false,"time":1720373428},"e4473cea9d5d13679d5320c8ffefdc18f793ae12":{"files":["CondTools/RunInfo/plugins/BuildFile.xml","CondTools/RunInfo/plugins/LHCInfoPerFillOnlinePopConAnalyzer.cc","CondTools/RunInfo/python/LHCInfoPerFillPopConAnalyzer_cfg.py","CondTools/RunInfo/python/LHCInfoPerLSPopConAnalyzer_cfg.py"],"squashed":false,"time":1720373433}},"emoji":{},"last_seen_sha":"54289f9f5218327afa66d532e75ea04360896045","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/40845

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
A new Pull Request was created by  @JanChyczynski for master.

It involves the following packages:

- CondTools/RunInfo (**db**)


@cmsbuild, @consuegs, @francescobrivio, @perrotta, @saumyaphor4252 can you please review it and eventually sign? Thanks.
@JanChyczynski, @PonIlya, @mmusich, @rsreds, @yuanchao this is something you requested to watch as well.
@antoniovilela, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42059

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

JanChyczynski committed:
Extract SourceHandler from LHCInfoPer*PopConAnalyzer

JanChyczynski committed:
For perLS duringfill implement adding defaultPayload and skipping invalid payloads

implement adding defaultPayload in duringFill mode

duringFill perLS: skip upload if invalid values

 Adjust duringFill unit tests

remove duplicated kLumisectionsQueryLimit

fix unit tests, print logs of failed lhcInfoNewPopCon test

add connection strings as cfg parameters, fix frontier key parameter for perFill

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42398

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

JanChyczynski commented:
The PR is now ready for tests and review. Please not that it depends on #44927 so it **has to be tested on top of it** and merged after it.

perrotta commented:
test parameters:
- pull_request = #44927

perrotta commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 40KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42431/summary.html
**COMMIT**: 7a1ca2d86e3bdaf0294dfc561c51aabc74530417
**CMSSW**: CMSSW_14_2_X_2024-10-28-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45395/42431/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

perrotta commented on CondTools/RunInfo/src/LHCInfoPerFillPopConSourceHandler.cc:
```diff
@@ -0,0 +1,746 @@
+
+std::string LHCInfoPerFillPopConSourceHandler::id() const { return m_name; }
+
+static constexpr unsigned int kLumisectionsQueryLimit = 4000;
```
It is already in the header file
```suggestion
```
But it would be even better if you remove this `constexpr` also from all the header files and use it in the cc's via `cond::lhcInfoHelper::kLumisectionsQueryLimit`, thus centralizing the definition (you are including the header of that helper anyhow)

JanChyczynski replied:
done, I've centralized the definition

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42402

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42404

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Size**: This PR adds an extra 44KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42441/summary.html
**COMMIT**: a83507271a52d895e65d77853db8650100704033
**CMSSW**: CMSSW_14_2_X_2024-10-29-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45395/42441/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 2 errors in the following unit tests:

<pre>---> test CondToolsLHCInfoTest had ERRORS
---> test CondToolsLHCInfoNewPopConTest had ERRORS
</pre>

## RelVals-INPUT

- [2024.000001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42441/runTheMatrixINPUT-results/2024.000001_RunJetMET02024D_10k)```2024.000001_RunJetMET02024D_10k/step1_dasquery.log```
- [2024.001001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42441/runTheMatrixINPUT-results/2024.001001_RunZeroBias2024D_10k)```2024.001001_RunZeroBias2024D_10k/step1_dasquery.log```
- [2024.100001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42441/runTheMatrixINPUT-results/2024.100001_RunJetMET02024C_10k)```2024.100001_RunJetMET02024C_10k/step1_dasquery.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [2024.101001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42441/runTheMatrixINPUT-results/2024.101001_RunZeroBias2024C_10k)
- 2024.000001
- 2024.001001
- 2024.100001
- 2024.101001
</details>



## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-29-1100+4f1f58/65334/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3569372
 - DQMHistoTests: [Total failures: 419](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-29-1100+4f1f58/65334/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3568933
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 201 log files, 171 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-10-29-1100+4f1f58/65334/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42439

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

JanChyczynski commented:
I believe the relval failures are unrelated (failing in other recent PRs as well). I've fixed one unit test (sorry for overseeing it) and I've attempted to fix the other but I'm only guessing what's the problem - I cannot reproduce it even in the test environment created with the command provided by the cmsBot. The logs don't help - not everything is in the output of the test script (I've fixed it now so all the info should be there if something fails). Maybe it's dependent on the path where you run the test and maybe my assumption that it's the CMSSW_.../src is wrong. I've made the test work regardless of the directory it's run in.

perrotta commented:
please test

JanChyczynski commented:
14_1 Backport open: https://github.com/cms-sw/cmssw/pull/46572

perrotta commented:
ping bot
(tests seem to be stuck)

perrotta commented:
please abort

perrotta commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 12KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42539/summary.html
**COMMIT**: 25d3a7fbba86e10bb77647a45bf74de74243d3a7
**CMSSW**: CMSSW_14_2_X_2024-11-01-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45395/42539/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 7 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-01-1100+4f1f58/65418/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3569567
 - DQMHistoTests: [Total failures: 528](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-01-1100+4f1f58/65418/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3569019
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 201 log files, 171 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-01-1100+4f1f58/65418/triggerResults): no differences found

perrotta commented:
+db
- It depends on https://github.com/cms-sw/cmssw/pull/44927, which is already merged

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

JanChyczynski commented:
I'm temporarily adding [Do not merge] to the title as I've just realized some additional changes are necessary

perrotta commented:
> I'm temporarily adding [Do not merge] to the title as I've just realized some additional changes are necessary

Removed db signature while waiting for the updates

JanChyczynski commented:
It's ready now @perrotta

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45395/42502

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CondTools/RunInfo/plugins/BuildFile.xml  modified in PR(s): #42841

cmsbuild commented:
Pull request #45395 was updated. @atpathak, @cmsbuild, @consuegs, @francescobrivio, @perrotta can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4f1f58/42598/summary.html
**COMMIT**: 6f55d49e303ccff02261c002cdb7795ca7c99521
**CMSSW**: CMSSW_14_2_X_2024-11-05-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45395/42598/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-05-1100+4f1f58/65463/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-05-1100+4f1f58/65463/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3343138
 - DQMHistoTests: [Total failures: 374](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-05-1100+4f1f58/65463/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3342744
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 195 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-11-05-1100+4f1f58/65463/triggerResults): no differences found

perrotta commented:
+db

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
